[1LP][RFR] Fix SSH timeout issue in verify_revert_snapshot

prichard77 opened the pull request:
## Purpose or Intent


- __Fixing__ verify_revert_snapshot. We were initializing the variables needed for ssh before the VM had IPs. This was causing failures on SSH. Usage of connect_ssh was added to fix this issue. 


### PRT Run

{{ pytest: --use-provider rhv43 --use-provider vsphere67-nested --long-running cfme/tests/infrastructure/test_snapshot.py::test_verify_revert_snapshot }}

#### PR failures
The test sometimes failed for reasons unrelated to the changes (timeout when waiting for `ap` command prompt) (see PRT runs)

john-dupuy commented on cfme/tests/infrastructure/test_snapshot.py:
```diff
@@ -211,9 +212,10 @@ def verify_revert_snapshot(full_test_vm, provider, soft_assert, register_event,
     else:
         snapshot1 = new_snapshot(full_test_vm)
     full_template = getattr(provider.data.templates, 'full_template')
+    ssh_ip = wait_pingable(full_test_vm.mgmt, wait=300, allow_ipv6=True)
```
You mentioned that this will sometimes get the local IP vs the correct IP, do you want to add a `wait_for` to add that check in?

prichard77 replied:
I believe wait_pingable handles that case. It looks for the pingable IP and the 192.x.x.x IP is not pingable AFAIK. This is the behavior I have seen during testing.

prichard77 replied:
I have seen the SSH piece pass in several test runs so far.

john-dupuy replied:
Ah yes, of course! Thanks @prichard77!

jaryn replied:
Waiting for machine being pingable is not good. What should be done is to make ssh trying to repeat the connecting attempt until timeout or until number of attempts exhausted.

The machine can become "pingable" but the service may not be started yet. So this has to be handled in the SSH connection-making code anyway, rendering the `wait_pingable` bit redundant. 

I did some work on the SSH client part about this alredy. I wonder what is the problem. I will take deeper look tomorrow.

prichard77 replied:
We do repeat the SSH attempt and handle the exception until timeout. We added the wait_pingable  to wait until we have a pingable IP to populate the SSH arguments. Previously, the hostname in ssh_kwargs was getting populated with None and it would always fail.
I believe the error in PRT we are seeing now may be an intermittent issue I have seen. It didn't show up while testing this locally today. I'll look into it more now.

jaryn replied:
https://github.com/ManageIQ/integration_tests/blob/master/cfme/fixtures/utility_vm.py#L44
solves a problem of picking the address on which the SSH port is responding and where the `true` command succeeds.

Perhaps this method should be somehow integrated in the ssh client machinery in our code base. There is some logic already and I find it bit intricate IIRC and not sure we have someone brave enough to poke in it at the moment. But we can simply add it and let the rest be.

Note that sometimes I see a "connection reset" on ssh connections when the CFME vm is being configured. I have no idea what is happening. Running SSH connections seem to withstand restarting of SSH daemon, so perhaps some firewall drops the packets.  The linked method of course cannot help with this situation.

prichard77 replied:
The code you linked above does look better than what was implemented in the legacy testcase initially. I am considering using it, but if ping will not work then ssh certainly will not work. The test will still timeout trying SSH. There seems to be an underlying issue with the containers or something in PRT as I can run this just fine locally (It fails after SSH for an issue with a BZ opened against it). I am a bit baffled. I wonder if vsphere providers have this issue on PRT? Locally, the wait_pingable was not needed for vsphere providers.

jaryn replied:
@prichard77 Not quite. In the past I found that the `ping` executable is not present in the PRT image. The code executing the ping did bad job reporting what went wrong so it looked like ping didn't pass trough though the real problem was that ping command was not found.

`ping` is only good to test whether the target OS is alive, but that is about all you can infer. It does no guarantee the required service is running, nor whether the service listens on the address you were pinging. It may listen on IPv6 only for example.

john-dupuy commented:
PR looks good, but PRT is failing with
```
        )[0]
E       wait_for.TimedOutError: Could not do 'function is_reachable()' at /cfme/cfme_tests/cfme/utils/net.py:213 in time

cfme/utils/net.py:227: TimedOutError
TimedOutError
b"Could not do 'function is_reachable()' at /cfme/cfme_tests/cfme/utils/net.py:213 in time"
```

prichard77 commented:
I believe the error in PRT we are seeing now may be an intermittent issue I have seen. It didn't show up while testing this locally today. I'll look into it more now.

prichard77 commented:
After further thought, PRT might be using the wrong template. I am looking in logs to see if I can determine this. We may be caught between issues. I have PR#9942 which fixes the isue with using the wrong template. I have the template updated to the correct value in cfme_data.yaml on my local machine. I am not 100% positive this is what is happening, but it is likely. We'll need to push PR9942 to get this to work (at least until it hits the other issue with active VM).

prichard77 commented:
I added vsphere provider to the PRT command to see what it does in PRT. It has been more stable locally. I am also testing the connect_ssh code that has been suggested, although I suspect it will also have issues..

dajoRH commented:
Lint report for commit d96c6c8b505c82a90152f3000dba80872ffcbf85:

`cfme/tests/infrastructure/test_snapshot.py`:
- :red_circle: Line 20:1: **F401** *'cfme.utils.net.wait_pingable' imported but unused*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

jaryn commented on cfme/tests/infrastructure/test_snapshot.py:
```diff
@@ -211,20 +211,14 @@ def verify_revert_snapshot(full_test_vm, provider, soft_assert, register_event,
+        'password': credentials[full_template.creds]['password']}
+    with connect_ssh(full_test_vm, creds) as ssh_client:
+        yield ssh_client
+    wait_for(lambda: ssh_client.run_command('touch snapshot1.txt').success, num_sec=200,
```
I think you should *not* yield here, but rather put the rest of the code until the line 265 under the `with` statement and remove the `ssh_client.close()` as the `with` will do that for you

prichard77 commented:
I was sure I updated the title of this PR when I updated the details, but I don;t see the change. I'll update it again.

prichard77 commented:
I am still testing and not seeing things work well 100% of the time. I'll keep working on it and I need to look at the suggestions from Jaroslov. I am pushing my updates anyway so we can discuss a few things in while I continue to debug and test. It also seems like a waste of time to need to wait for the "wrong" IP to time out. Is thee another way besides just checking to see if we have the 10.x.x.x IP versus the 192.x.x.x IP? I am considering checking the IP with a variable like ip_class_octet so we can update if the network settings change. Possibly get this from yaml? suggestions?

prichard77 commented:
I'll fix the flake8 needed balank line in next commit. Ignore for now. there should be more changes.

dajoRH commented:
Lint report for commit 640890b05c00a2ad3719124bca2026be61e35b37:

`cfme/fixtures/utility_vm.py`:
- :bangbang: Line 46:38: **E261** *at least two spaces before inline comment*
- :bangbang: Line 48:101: **E501** *line too long (103 > 100 characters)*

`cfme/tests/infrastructure/test_snapshot.py`:
- :bangbang: Line 208:1: **E302** *expected 2 blank lines, found 1*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

prichard77 commented:
I added a loop and wait, and also a check for the class octet of IPs found for the VM to connect_ssh. I believe this fixes the ssh issue with out test case, but now if seems like I have "hijacked' utility code that was intended to be reusable and have made it pretty specific to the needs of the test case. I am also curious about the AttrDict I added. Comments?

prichard77 commented:
I took a look at Jaroslav's suggestions and have answered some concerns. Still more to work out. I liked his usage of creds = SSHCredential.from_config(full_template.creds). I was considering something like this, but hadn't spent the time to find or write one. I'll add that. Then I won't need AttrDict.
Have we made connect_ssh too specific (with my addition of extra loop and octet check) to be a util. Should I make it a function somewhere? in test_snapshot.py? (leave the original but put my updated copy somewhere else)?   
I couldn't find any usage of connect_ssh, so I don't know what it was originally intended for?

jaryn commented on cfme/fixtures/utility_vm.py:
```diff
@@ -32,29 +32,40 @@
-                return
-        except Exception as ex:
-            logger.info("Failed to connect with SSH to %s: %s", ip, ex)
+        for ssh_attempt in range(55):  # I know ... hardcoded. I'll fix
```
Not good as the _trying_ips already does 60 attempts by default

prichard77 replied:
Doesn't the _trying_ips just do 60 attempts to get an IP? Once we get one, we need to try to connect via SSH until the SSH client is up (until we are successful).

jaryn commented on cfme/fixtures/utility_vm.py:
```diff
@@ -32,29 +32,40 @@
+                    return
+            except Exception as ex:
+                logger.info("Failed to connect with SSH to %s: %s", ip, ex)
+                time.sleep(20)
```
_trying_ips is sleeping after trying all the different IPs returned from the mgmt system. I wonder why do you want it to sleep after every unsuccessful connection.

prichard77 replied:
To wait for the SSH client? We keep going around in circles about this. Am I missing something? We wait until we get IP,  then we need to wait for the SSH client to become active. I have been debugging and AFAICT once, we get an IP, we only try to get an SSH connection once. Am I wrong? I am running this code locally and I am unable to get it to pass. SSH is timing out. Maybe the difference between running locally and running in PRT is causing some of this confusion. I'd really like to be able to run this test in both places.

jaryn commented on cfme/fixtures/utility_vm.py:
```diff
@@ -32,29 +32,40 @@
-            yield ip
-            return
+            try:
+                if ip.split('.')[0] == '10':
```
This is attempt to resolve problem of too-long timeout when IP is not responding. This approach assumes we are dealing with system in RedHat network thus not quite as robust as if this problem was dealt with by passing some sort of `connection_timeout` argument that would be used to set the Paramiko's `timeout` of the `connect` method: http://docs.paramiko.org/en/2.4/api/client.html#paramiko.client.SSHClient.connect

prichard77 replied:
This is an attempt to solve the issue of the 192 IPs NEVER allowing an SSH connection wrt this test. We must wait for the attempts to time out. We must sometimes wait a few minutes for the SHH client to come up. If we get the 192 IP first, we'll need to wait for the entire timeout.  I agree this could be handled by increasing the timeout in paramiko and we wouldn't need an additional sleep loop, but we'd still be waiting for 192 to time out.

jaryn replied:
You are right the 192.x address makes delay. But I think you are missing that all ips are meant to be tried in a short while (X), then wait and then again:
```
all_ips returns no IPs .. nothing to do
wait X seconds
all_ips returns no IPs .. nothing to do
wait X seconds
all_ips returns [192.x.y.z] 
   .. attempt to connect 192.x.y.z... timeout after T seconds
wait X secodes
all_ips returns [192.x.y.z, IP_somewhere_in_the_universe]... 
   .. attempt to connect 192.x.y.z, timeout after T seconds
   .. atempt to connect  IP_somewhere_in_the_universe .. perhaps timeout after T seconds as the service is not responding or something strange happening (we will most probably get Connection refused  immediately once the OS is up)
wait X seconds
all_ips returns [192.x.y.z, IP_somewhere_in_the_universe]...
   .. attempt to connect 192.x.y.z, timeout after T seconds
   .. atempt to connect  IP_somewhere_in_the_universe .. connects as service become available
```

Yeah, with more addresses that are time-outing on connections, our delays to try them all become longer. But the timeout (T) for connection attempt doesn't need to be like 60 seconds or whatever is the default but something like several seconds.

jaryn commented:
> Have we made connect_ssh too specific (with my addition of extra loop and octet check) to be a util. Should I make it a function somewhere? in test_snapshot.py? (leave the original but put my updated copy somewhere else)?

I think the `connect_ssh` should be moved to perhaps the SSHClient as long as it is kept working with any IP. I think if some more though is put in, it can be named somehow better to make its purpose more obvious. It could also probably be made accepting some closure which would be used for obtaining the list of IPs for every trials round. This would make it useful for other objects, not just mgmt VMs. I think none of the work stated in this paragraph needs to be done right now in this PR or the #9956.

> I couldn't find any usage of connect_ssh, so I don't know what it was originally intended for?

I don't recall what exactly I was using it for, but it really was for situation where the target VM was being booted and I had to make several attempts until it was ready to accept SSH connections. This piece of code become dead code by some with my other changes I did later in the same PR.

prichard77 commented:
It's worth noting that I am planning on implementing some sort of ssh_connection property for the Base_vm Class. This would be a cached property. I was going to get the test working with connect_ssh and then in another PR I'd add the property to Base_vm. I was thinking we were close to a solution here so I wanted to get this PR merged before I started work on the property, but I guess not

prichard77 commented:
Well the last PRT run passed. I am skeptical but hopeful. Maybe the test passes more consistently on PRT? I still need to do more testing as I have seen false positives before. And I need to rebase.

dajoRH commented:
Would you mind rebasing this Pull Request against latest master, please? :trollface:
*CFME QE Bot*

dajoRH commented:
Lint report for commit 50a7b4a91c168e8aedac5275257b96080994295e:

`cfme/tests/test_framework.py`:
- :heavy_exclamation_mark: Line 49:19: **W292** *no newline at end of file*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

prichard77 commented:
I am unable to get the unit tests working locally. Still working on it. `

> def nosleep(mocker):
> E       fixture 'mocker' not found`

dajoRH commented:
I detected some fixture changes in commit 50a7b4a91c168e8aedac5275257b96080994295e

The *local* fixture **`nosleep`** is used in the following files:
- **cfme/tests/test_framework.py**

The *local* fixture **`vm_mock`** is used in the following files:
- **cfme/tests/test_framework.py**
    - *test_connect_ssh*

Please, consider creating a PRT run to make sure your fixture changes do not break existing usage :smiley:

prichard77 commented:
Any comments/suggestions on where to put the unit tests? I put them alone in a file named test_connect_ssh.py. Do we want to keep ssh stuff together or leave as is here?

dajoRH commented:
I detected some fixture changes in commit c93f63dacaf7307a797a7c001e30fae6484cc33c

The *local* fixture **`nosleep`** is used in the following files:
- **cfme/utils/tests/test_connect_ssh.py**

The *local* fixture **`vm_mock`** is used in the following files:
- **cfme/utils/tests/test_connect_ssh.py**
    - *test_connect_ssh*

Please, consider creating a PRT run to make sure your fixture changes do not break existing usage :smiley:

jaryn commented:
> Any comments/suggestions on where to put the unit tests? I put them alone in a file named test_connect_ssh.py. Do we want to keep ssh stuff together or leave as is here?
I think you need to add 
`pytest-mock==2.0.0`
to the `requirements/frozen.txt`

prichard77 commented:
frozen.txt is updated. It's at the bottom of the view when looking at changes files.

prichard77 commented:
I just pushed some updates to the unit tests for connect_ssh. I added a case where the first IP returned is WORKING. There may be a more eloquent way of coding this, but I wanted to get the idea across without rewriting the entire test. Let me know what you think. Otherwise I think this is good to go for now. We can certainly tweak as needed.

dajoRH commented:
I detected some fixture changes in commit 84bc6d0f1a020e12e31500199e9bdb1acd3830dc

The *local* fixture **`nosleep`** is used in the following files:
- **cfme/utils/tests/test_connect_ssh.py**

The *local* fixture **`vm_mock`** is used in the following files:
- **cfme/utils/tests/test_connect_ssh.py**
    - *test_connect_ssh*

The *local* fixture **`vm_mock2`** is used in the following files:
- **cfme/utils/tests/test_connect_ssh.py**
    - *test_connect_ssh*

Please, consider creating a PRT run to make sure your fixture changes do not break existing usage :smiley:

jaryn commented on cfme/utils/tests/test_connect_ssh.py:
```diff
@@ -25,7 +25,19 @@ def vm_mock():
+    all_ips_mock = PropertyMock()
+    type(vm).all_ips = all_ips_mock
+    all_ips_mock.side_effect = [
+        ['WORKING_IP_MOCK', 'SHOULD_NOT_REACH_THIS_MOCK']
```
```suggestion
        'WORKING_IP_MOCK', 'SHOULD_NOT_REACH_THIS_MOCK'],
        ['SHOULD_NOT_REACH_THIS_MOCK', 'SHOULD_NOT_REACH_THIS_MOCK']
    ]
    return vm


@pytest.fixture
def vm_mock2():
    vm = Mock()
    all_ips_mock = PropertyMock()
    type(vm).all_ips = all_ips_mock
    all_ips_mock.side_effect = [
        ['WORKING_IP_MOCK', 'SHOULD_NOT_REACH_THIS_MOCK']

```

prichard77 replied:
I updated as suggested, I believe.

prichard77 commented on cfme/fixtures/utility_vm.py:
```diff
@@ -34,7 +34,7 @@
 def _trying_ips(vm, attempts=60, interval=10):
     for attempt in range(attempts):
-        for ip in set(vm.all_ips):
+        for ip in getattr(vm, 'all_ips', []):
```
Wouldn't that cause an exception to get raised when we tried to get IPs before they exist on the VM? The [] allows for something to get returned without an exception when we have no IPs and then we try again.

prichard77 replied:
I am testing because i am not 100% sure.

jaryn replied:
> Wouldn't that cause an exception to get raised when we tried to get IPs before they exist on the VM? The [] allows for something to get returned without an exception when we have no IPs and then we try again.

In such case I would expect the property to equal to empty iterable (A `[]` for example).  I wouldn't expect it to not exist and thus to raise an exception.

jaryn replied:
@prichard77 https://github.com/RedHatQE/wrapanapi/blob/master/wrapanapi/entities/vm.py#L164

This means all descendant should have this property defined and `gettattr` is not useful. It won't change a thing except of obscuring:
```
In [5]: class Abstract(metaclass=abc.ABCMeta): 
   ...:     @abc.abstractproperty 
   ...:     def foo(self): 
   ...:         pass 
   ...:                                                                                       

In [6]: class Derived(Abstract): 
   ...:     pass 
   ...:                                                                                       

In [7]: Derived().foo                                                                         
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-7-5d49a38ba127> in <module>
----> 1 Derived().foo

TypeError: Can't instantiate abstract class Derived with abstract methods foo

In [8]: getattr(Derived(), 'foo')                                                             
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-8-58fdc29ee537> in <module>
----> 1 getattr(Derived(), 'foo')

TypeError: Can't instantiate abstract class Derived with abstract methods foo
```

prichard77 replied:
OK. vm.all_ips looks good. Added as suggested.

dajoRH commented:
I detected some fixture changes in commit 2273966a5a3fd6c4751adf261087221ee1d4e616

The *local* fixture **`nosleep`** is used in the following files:
- **cfme/utils/tests/test_connect_ssh.py**

The *local* fixture **`vm_mock`** is used in the following files:
- **cfme/utils/tests/test_connect_ssh.py**
    - *test_connect_ssh*

The *local* fixture **`vm_mock2`** is used in the following files:
- **cfme/utils/tests/test_connect_ssh.py**
    - *test_connect_ssh*

Please, consider creating a PRT run to make sure your fixture changes do not break existing usage :smiley:

john-dupuy commented on cfme/tests/infrastructure/test_snapshot.py:
```diff
@@ -185,25 +185,19 @@ def verify_revert_snapshot(full_test_vm, provider, soft_assert, register_event,
+    # connect_ssh will iterate over "all_ips" on the VM and return a client when it can connect
+
+    with connect_ssh(full_test_vm.mgmt, creds) as ssh_client:
+        # put wait for here.
```
Can this comment be removed?

prichard77 commented:
snapshot.active in virtual_machines.py still needs some work. I updated it so verify_revert_snapshot will work, but snapshot.active is very specific to the test and cannot be used generically. We can get comments on the other updates while I work on a solution to snapshot.active.

dajoRH commented:
I detected some fixture changes in commit c86dce2a34e260d4614438bfcacc09b569d59740

The *global* fixture **`utility_vm_nfs_ip`** is used in the following files:
- **cfme/tests/cli/test_appliance_console_db_restore.py**

The *global* fixture **`utility_vm_samba_ip`** is used in the following files:
- **cfme/tests/cli/test_appliance_console_db_restore.py**

The *global* fixture **`utility_vm_proxy_data`** is used in the following files:
- **cfme/tests/configure/test_proxy.py**
    - *prepare_proxy_specific*
    - *prepare_proxy_default*

The *global* fixture **`utility_vm_ssh`** is used in the following files:
- **cfme/tests/configure/test_proxy.py**
    - *validate_proxy_logs*
    - *prepare_proxy_specific*
    - *prepare_proxy_default*
    - *test_proxy_valid*
    - *test_proxy_override*

The *local* fixture **`nosleep`** is used in the following files:
- **cfme/utils/tests/test_connect_ssh.py**

The *local* fixture **`vm_mock`** is used in the following files:
- **cfme/utils/tests/test_connect_ssh.py**
    - *test_connect_ssh*

The *local* fixture **`vm_mock2`** is used in the following files:
- **cfme/utils/tests/test_connect_ssh.py**
    - *test_connect_ssh*

Please, consider creating a PRT run to make sure your fixture changes do not break existing usage :smiley:

dajoRH commented:
I detected some fixture changes in commit 4465520f3235968782d7379ecf577cc95b14fbb2

The *global* fixture **`utility_vm_nfs_ip`** is used in the following files:
- **cfme/tests/cli/test_appliance_console_db_restore.py**

The *global* fixture **`utility_vm_samba_ip`** is used in the following files:
- **cfme/tests/cli/test_appliance_console_db_restore.py**

The *global* fixture **`utility_vm_proxy_data`** is used in the following files:
- **cfme/tests/configure/test_proxy.py**
    - *prepare_proxy_specific*
    - *prepare_proxy_default*

The *global* fixture **`utility_vm_ssh`** is used in the following files:
- **cfme/tests/configure/test_proxy.py**
    - *validate_proxy_logs*
    - *prepare_proxy_specific*
    - *prepare_proxy_default*
    - *test_proxy_valid*
    - *test_proxy_override*

The *local* fixture **`nosleep`** is used in the following files:
- **cfme/utils/tests/test_connect_ssh.py**

The *local* fixture **`vm_mock`** is used in the following files:
- **cfme/utils/tests/test_connect_ssh.py**
    - *test_connect_ssh*

The *local* fixture **`vm_mock2`** is used in the following files:
- **cfme/utils/tests/test_connect_ssh.py**
    - *test_connect_ssh*

Please, consider creating a PRT run to make sure your fixture changes do not break existing usage :smiley:

prichard77 commented:
The failure of verify_revert_snapshot is now related to BZ1805803. The ssh is working, but we timeout waiting for the test to detect that the snapshot is active. I added a blocker marker, but the ssh portion has been tested pretty well this week.

dajoRH commented:
Lint report for commit 974c4c4747dada5749d2f355eb5c5e8416848171:

`cfme/tests/infrastructure/test_snapshot.py`:
- :bangbang: Line 181:1: **E302** *expected 2 blank lines, found 1*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

dajoRH commented:
I detected some fixture changes in commit 039e4f0a5fb86e89fa00f41a5b0bdf7bba871204

The *global* fixture **`utility_vm_nfs_ip`** is used in the following files:
- **cfme/tests/cli/test_appliance_console_db_restore.py**

The *global* fixture **`utility_vm_samba_ip`** is used in the following files:
- **cfme/tests/cli/test_appliance_console_db_restore.py**

The *global* fixture **`utility_vm_proxy_data`** is used in the following files:
- **cfme/tests/configure/test_proxy.py**
    - *prepare_proxy_specific*
    - *prepare_proxy_default*

The *global* fixture **`utility_vm_ssh`** is used in the following files:
- **cfme/tests/configure/test_proxy.py**
    - *validate_proxy_logs*
    - *prepare_proxy_specific*
    - *prepare_proxy_default*
    - *test_proxy_valid*
    - *test_proxy_override*

The *local* fixture **`nosleep`** is used in the following files:
- **cfme/utils/tests/test_connect_ssh.py**

The *local* fixture **`vm_mock`** is used in the following files:
- **cfme/utils/tests/test_connect_ssh.py**
    - *test_connect_ssh*

The *local* fixture **`vm_mock2`** is used in the following files:
- **cfme/utils/tests/test_connect_ssh.py**
    - *test_connect_ssh*

Please, consider creating a PRT run to make sure your fixture changes do not break existing usage :smiley:

prichard77 commented:
I see I need to add some more tests to PRT. But otherwise, how do we want to proceed. You can see in the PRT output that the ssh piece is not failing, but the part covered by the BZ is.

jaryn commented on cfme/tests/infrastructure/test_snapshot.py:
```diff
@@ -177,33 +179,28 @@ def test_delete_all_snapshots(create_vm, provider):
              message="Waiting for second snapshot to disappear")
 
 
+@pytest.mark.meta(blockers=[BZ(1805803, forced_streams=["5.10"])], automates=[1805803])
```
@prichard77 I think this should prevent the test to be ran with the failing configuration.  Did you try find out why it is not effective?

prichard77 replied:
I just thought that the blocker line was not effective when you specifically entered it in the PRT section of the PR. It shouldn't run when we do a full collection and run of tests. I am actually glad it runs here so we can check out the failures and continue with review.

prichard77 replied:
But the blocker was in the wrong place. I have put it in the correct place now. We can look at the old runs but it will get skipped in 5.10. BZ 1805803 is in 5.11 but it does not fix the issue.I have put it back to on_Dev. Thanks for pointing this issue out, Jaroslav.

prichard77 commented:
OH wait. I see I put the blocker in the wrong place. Fixing.

dajoRH commented:
I detected some fixture changes in commit 9fd14a386031e89b23553ae70fccd2cebb273382

The *global* fixture **`utility_vm_nfs_ip`** is used in the following files:
- **cfme/tests/cli/test_appliance_console_db_restore.py**

The *global* fixture **`utility_vm_samba_ip`** is used in the following files:
- **cfme/tests/cli/test_appliance_console_db_restore.py**

The *global* fixture **`utility_vm_proxy_data`** is used in the following files:
- **cfme/tests/configure/test_proxy.py**
    - *prepare_proxy_specific*
    - *prepare_proxy_default*

The *global* fixture **`utility_vm_ssh`** is used in the following files:
- **cfme/tests/configure/test_proxy.py**
    - *validate_proxy_logs*
    - *prepare_proxy_specific*
    - *prepare_proxy_default*
    - *test_proxy_valid*
    - *test_proxy_override*

The *local* fixture **`nosleep`** is used in the following files:
- **cfme/utils/tests/test_connect_ssh.py**

The *local* fixture **`vm_mock`** is used in the following files:
- **cfme/utils/tests/test_connect_ssh.py**
    - *test_connect_ssh*

The *local* fixture **`vm_mock2`** is used in the following files:
- **cfme/utils/tests/test_connect_ssh.py**
    - *test_connect_ssh*

Please, consider creating a PRT run to make sure your fixture changes do not break existing usage :smiley:

john-dupuy commented on cfme/fixtures/utility_vm.py:
```diff
@@ -17,77 +17,50 @@
-    else:
-        raise TimedOutError(f"Coudln't find an IP of vm {vm} with port {port} responding")
+ATTEMPT_TIMEOUT = 10
+""" How long to wait until connection is determined as not successful. """
```
```suggestion
# How long to wait until connection is determined as not successful.
```

prichard77 replied:
Thanks. The SSH code was written by @jaryn Jarsolav. Is this the reason you removed the 1LR?

john-dupuy commented:
Failures on PRT do not seem to be related to this PR.

prichard77 commented:
What was the reason for removing [1LP]?

john-dupuy commented:
Waiting on @jaryn to give his approval before moving to 1LP @prichard77

jaryn commented:
Good!

jaryn commented on cfme/tests/infrastructure/test_snapshot.py:
```diff
@@ -217,35 +213,28 @@ def verify_revert_snapshot(full_test_vm, provider, soft_assert, register_event,
 
 
 @pytest.mark.rhv1
+@pytest.mark.meta(blockers=[BZ(1805803, forced_streams=["5.10"])], automates=[1805803])
```
@PJ Richardson I see we have both providers blocked by BZ blocker as is.

I think the problem is only with one of them and only with CFME 5.10.

I am not sure is there not a better way, but the most flexible way to "block" is having this somewhere (perhaps on top) of the test method:
```
if provider.one_of(RHEVProvider) and appliance.version.stream = "5.10":
  pytest.skip()
```

Note I didn't check the above code has no bugs, but I think you got the idea.

prichard77 replied:
First, this issue is in both 5.10 and 5.11. It's not consistent so you may get lucky and see some passing instances. Your suggestion would skip the rhv tests. There are a couple ways we can accomplish this functionality. As above, with an uncollect_if, or by changing the providers used for this test. My concern is that if we remove the blocker market, we won't have the metric data reflect tests blocked by the BZ. Unless I can tie in the BZ to  pytest.skip(). I'll get some input from the team.

jaryn commented:
I think it would be good to let run what can run and pass with a conditional skip

dajoRH commented:
I detected some fixture changes in commit bb6b87295f5d40a4226563b246aa60c1f69822fe

The *global* fixture **`utility_vm_nfs_ip`** is used in the following files:
- **cfme/tests/cli/test_appliance_console_db_restore.py**

The *global* fixture **`utility_vm_samba_ip`** is used in the following files:
- **cfme/tests/cli/test_appliance_console_db_restore.py**

The *global* fixture **`utility_vm_proxy_data`** is used in the following files:
- **cfme/tests/configure/test_proxy.py**
    - *prepare_proxy_specific*
    - *prepare_proxy_default*

The *global* fixture **`utility_vm_ssh`** is used in the following files:
- **cfme/tests/configure/test_proxy.py**
    - *validate_proxy_logs*
    - *prepare_proxy_specific*
    - *prepare_proxy_default*
    - *test_proxy_valid*
    - *test_proxy_override*

The *local* fixture **`nosleep`** is used in the following files:
- **cfme/utils/tests/test_connect_ssh.py**

The *local* fixture **`vm_mock`** is used in the following files:
- **cfme/utils/tests/test_connect_ssh.py**
    - *test_connect_ssh*

The *local* fixture **`vm_mock2`** is used in the following files:
- **cfme/utils/tests/test_connect_ssh.py**
    - *test_connect_ssh*

Please, consider creating a PRT run to make sure your fixture changes do not break existing usage :smiley:

john-dupuy commented:
looks like it's passing now @jaryn :smile:

prichard77 commented:
I am not sure why Travis CI is showing - Pending. When I look at it is looks all green and my last push was over a day ago.

mshriver commented on cfme/utils/net.py:
```diff
@@ -58,6 +60,39 @@ def ip_echo_socket(port=32123):
+        time.sleep(delay)
+
+
+def pick_responding_ip(vm, port, rounds, rounds_delay_seconds, attempt_timeout):
```
Optional, future addition - 

I think we can likely default this to 22, assuming we're looking for something that is responding for SSH.

jaryn replied:
I would like it to stay as it is. While I see this function could be used for finding IP connective with SSH in the future quite a lot, It already is used for finding IPs responding on other than SSH ports (NFS and CIFS). It is meant to be generic. For same reason it is put in `cfme.utils.net`.

mshriver commented on cfme/utils/net.py:
```diff
@@ -58,6 +60,39 @@ def ip_echo_socket(port=32123):
             conn.close()
 
 
+def _trying_fresh_ips(vm, delay, rounds=-1):
```
I don't think waiting indefinitely should be the default behavior for this method.

I'm thinking a default of 3 rounds would be sufficient. Indefinite wait should be an explicit call.

jaryn replied:
Yeah, maybe it is not a best idea to make it go forever. Perhaps it should be required parameter as the number of rounds is tightly coupled with the delay (when you change one, you likely need to change the other as well).

john-dupuy replied:
@prichard77 what are your thoughts here?

jaryn replied:
I am making it a mandatory parameter. I am making it run infinitely when one will pass `math.inf` instead of -1. `math.inf` type is `float`, but I don't think it matters.

mshriver commented on cfme/utils/ssh.py:
```diff
@@ -842,3 +851,48 @@ def keygen():
+                               use_check_port=False, timeout=connect_timeout)
+            client.connect()
+        except Exception as ex:
+            logger.info(f"Failed to connect with SSH to {ip}: {ex}")
```
Please use `logger.exeception` here, as it will include the traceback context, not just the exception message.

prichard77 replied:
Updated as suggested.

mshriver commented on cfme/utils/ssh.py:
```diff
@@ -842,3 +851,48 @@ def keygen():
+            logger.info(f"SSH connected to IP {ip}")
+
+            # This command execution check is probably redundant.
+            result = client.run_command("true")
```
Agreed this is redundant, I'd remove it, no need for the extra load of making a token command call.

prichard77 replied:
Removed.

mshriver commented on requirements/frozen.txt:
```diff
@@ -259,6 +259,7 @@ pyperclip==1.7.0
 pytesseract==0.3.0
 pytest==5.2.1
 pytest-polarion-collect==0.23.1
+pytest-mock==2.0.0
```
This requirement needs to be in the template_non_imported.txt file, as its not directly imported but the fixture is indeed used in the unit test.

When adding requirements, this can be done by specifying
```
miq requirement add --extra pytest-mock
```

prichard77 replied:
I put the line in template_non_imported.txt as the last line. Let me know if it is the wrong place. Also, I need to keep line 262 above?

mshriver replied:
thanks, and correct, still needs to be in frozen.

the requirement script will install both template_scanned.txt and template_non_imported.txt, with the constraints file, and then generate this frozen file from the resulting venv packages.

dajoRH commented:
I detected some fixture changes in commit ae8a5b0aefec2a52196c302d7067d907ae392698

The *global* fixture **`utility_vm_nfs_ip`** is used in the following files:
- **cfme/tests/cli/test_appliance_console_db_restore.py**

The *global* fixture **`utility_vm_samba_ip`** is used in the following files:
- **cfme/tests/cli/test_appliance_console_db_restore.py**

The *global* fixture **`utility_vm_proxy_data`** is used in the following files:
- **cfme/tests/configure/test_proxy.py**
    - *prepare_proxy_specific*
    - *prepare_proxy_default*

The *global* fixture **`utility_vm_ssh`** is used in the following files:
- **cfme/tests/configure/test_proxy.py**
    - *validate_proxy_logs*
    - *prepare_proxy_specific*
    - *prepare_proxy_default*
    - *test_proxy_valid*
    - *test_proxy_override*

The *local* fixture **`nosleep`** is used in the following files:
- **cfme/utils/tests/test_connect_ssh.py**

The *local* fixture **`vm_mock`** is used in the following files:
- **cfme/utils/tests/test_connect_ssh.py**
    - *test_connect_ssh*

The *local* fixture **`vm_mock2`** is used in the following files:
- **cfme/utils/tests/test_connect_ssh.py**
    - *test_connect_ssh*

Please, consider creating a PRT run to make sure your fixture changes do not break existing usage :smiley:

dajoRH commented:
I detected some fixture changes in commit b216e26e1a482a28c1dc275bf307cce5d5985e10

The *global* fixture **`utility_vm_nfs_ip`** is used in the following files:
- **cfme/tests/cli/test_appliance_console_db_restore.py**

The *global* fixture **`utility_vm_samba_ip`** is used in the following files:
- **cfme/tests/cli/test_appliance_console_db_restore.py**

The *global* fixture **`utility_vm_proxy_data`** is used in the following files:
- **cfme/tests/configure/test_proxy.py**
    - *prepare_proxy_specific*
    - *prepare_proxy_default*

The *global* fixture **`utility_vm_ssh`** is used in the following files:
- **cfme/tests/configure/test_proxy.py**
    - *validate_proxy_logs*
    - *prepare_proxy_specific*
    - *prepare_proxy_default*
    - *test_proxy_valid*
    - *test_proxy_override*

The *local* fixture **`nosleep`** is used in the following files:
- **cfme/utils/tests/test_connect_ssh.py**

The *local* fixture **`vm_mock`** is used in the following files:
- **cfme/utils/tests/test_connect_ssh.py**
    - *test_connect_ssh*

The *local* fixture **`vm_mock2`** is used in the following files:
- **cfme/utils/tests/test_connect_ssh.py**
    - *test_connect_ssh*

Please, consider creating a PRT run to make sure your fixture changes do not break existing usage :smiley:

prichard77 commented:
Taking a look at your last comment, I believe it is related to work we are planning for ssh. We had discussed implementing a cached property that would return an ssh client for things like appliances and more. Is this what you were suggesting? I wanted to get the ssh functionality working, and get the updates to the test pushed before I started to work on the cached property.

prichard77 commented:
> I don't think my concern is significant enough to block this PR
Is there something else that caused the PR to get moved to WIPTEST from 1LP?

mshriver commented on requirements/template_non_imported.txt:
```diff
@@ -11,3 +11,4 @@ polarion-tools-common
 pre-commit
 pytest-polarion-collect
 pytest-report-parameters
+miq requirement add --extra pytest-mock
```
So, this is a command that you need to run on your CLI, which will modify this file, and add `pytest-mock` to it, sorted.

prichard77 replied:
I ran the command and pushed the update to requirements/template_non_imported.txt. This is the only file in my repo that was updated. You mention template_scanned.txt  above, but I'm assuming it didn't need to get updated for pytest-mock, but would have been if it was needed.

dajoRH commented:
I detected some fixture changes in commit 3ff036389c7f94cc0311adde7df0d07654b6802d

The *global* fixture **`utility_vm_nfs_ip`** is used in the following files:
- **cfme/tests/cli/test_appliance_console_db_restore.py**

The *global* fixture **`utility_vm_samba_ip`** is used in the following files:
- **cfme/tests/cli/test_appliance_console_db_restore.py**

The *global* fixture **`utility_vm_proxy_data`** is used in the following files:
- **cfme/tests/configure/test_proxy.py**
    - *prepare_proxy_specific*
    - *prepare_proxy_default*

The *global* fixture **`utility_vm_ssh`** is used in the following files:
- **cfme/tests/configure/test_proxy.py**
    - *validate_proxy_logs*
    - *prepare_proxy_specific*
    - *prepare_proxy_default*
    - *test_proxy_valid*
    - *test_proxy_override*

The *local* fixture **`nosleep`** is used in the following files:
- **cfme/utils/tests/test_connect_ssh.py**

The *local* fixture **`vm_mock`** is used in the following files:
- **cfme/utils/tests/test_connect_ssh.py**
    - *test_connect_ssh*

The *local* fixture **`vm_mock2`** is used in the following files:
- **cfme/utils/tests/test_connect_ssh.py**
    - *test_connect_ssh*

Please, consider creating a PRT run to make sure your fixture changes do not break existing usage :smiley:

prichard77 commented:
PRT is failing for a test that I added a blocker to. I am not sure why but I added this 

> @pytest.mark.meta(
>     blockers=[BZ(1805803, unblock=lambda provider: not provider.one_of(RHEVMProvider),
>                  forced_streams=["5.10"])], automates=[1805803])

to test_verify_revert_snapshot. The failing test is test_verify_revert_snapshot[rhevm-4.3-full_template]. I am not sure why it is getting run. Maybe a PRT issue?

mshriver commented on cfme/utils/ssh.py:
```diff
@@ -842,3 +851,42 @@ def keygen():
+    timestamp = time.monotonic()
+    # Note we do need to be sure we get the fresh list of IPs for every cycle.
+    # The provider may be updating it.
+    for ip in _trying_fresh_ips(vm, delay=delay, rounds=-1):
```
Why isn't this using `pick_responding_ip`, as that wraps this method anyway?

I am not a fan of the pattern here, setting up an infinite generator and then comparing timestamps. It serves almost the same exact purpose as a wait_for call.

A net check is also less costly than trying to establish the ssh connection immediately.

jaryn replied:
### Wait for
I am not sure `wait_for` could be used. When I started this effort, I didn't have the `use_check_port` attribute. Perhaps now when it is implemented wait_for could be used. I will think this trough.

### Net check
How is the `net_check` less costly? What kind of resources you are talking about? I think the opposite is true. Checking with `net_check` means you need to be creating the connection twice -- once for checking, the other time for the connection that will be used for payload. So why to not just trying to connect with some timeout and try again if it timeouts? Is it because of the SSHClient's object creation costs?

If server is connective first time, it doesn't need to be the case the second time (when the payload-serving connection attempt is made) as for example some lousy firewall might start meanwhile or some transition effect (perhaps services deps cycles could be causing this) can happen. Once we have connection established, it may survive such things (for example ssh will survive restart of the sshd as it is not killing it's children). And yes, issues like that are real. I think services deps issues may be the case why I am getting connection_reset in some of my tests. One such test is included in this PR and makes PRT failures sometimes. It may be great to have a wrapper around the SSHClient to make it reliably reconnecting. It is not working fine now even though I think there was some effort to make it reliable.

prichard77 replied:
I am making this change now, and testing. I just did a simple replacement of the function we call and added the args. However, I believe this undoes some of the ability to pass values down that was initially intended. This was written by Jaroslav so I need to take a closer look and possibly get him involved.

prichard77 replied:
Testing is failing so I am debugging.

prichard77 replied:
So how do we want to proceed on this? I really thought we had come up with an appropriate solution. Seems like we need to nail down what are the specific needs/requirement that we are trying to accomplish with these updates to connect_ssh.

dajoRH commented:
I detected some fixture changes in commit ecd6b6c909ecd7f960ee47bfcaa6799623ea937e

The *global* fixture **`utility_vm_nfs_ip`** is used in the following files:
- **cfme/tests/cli/test_appliance_console_db_restore.py**

The *global* fixture **`utility_vm_samba_ip`** is used in the following files:
- **cfme/tests/cli/test_appliance_console_db_restore.py**

The *global* fixture **`utility_vm_proxy_data`** is used in the following files:
- **cfme/tests/configure/test_proxy.py**
    - *prepare_proxy_specific*
    - *prepare_proxy_default*

The *global* fixture **`utility_vm_ssh`** is used in the following files:
- **cfme/tests/configure/test_proxy.py**
    - *validate_proxy_logs*
    - *prepare_proxy_specific*
    - *prepare_proxy_default*
    - *test_proxy_valid*
    - *test_proxy_override*

The *local* fixture **`nosleep`** is used in the following files:
- **cfme/utils/tests/test_connect_ssh.py**

The *local* fixture **`vm_mock`** is used in the following files:
- **cfme/utils/tests/test_connect_ssh.py**
    - *test_connect_ssh*

The *local* fixture **`vm_mock2`** is used in the following files:
- **cfme/utils/tests/test_connect_ssh.py**
    - *test_connect_ssh*

Please, consider creating a PRT run to make sure your fixture changes do not break existing usage :smiley:

dajoRH commented:
Would you mind rebasing this Pull Request against latest master, please? :trollface:
*CFME QE Bot*

dajoRH commented:
I detected some fixture changes in commit 2ccfd21c0c5d4b6dcd6c281f0671cc86831b9178

The *global* fixture **`utility_vm_nfs_ip`** is used in the following files:
- **cfme/tests/cli/test_appliance_console_db_restore.py**

The *global* fixture **`utility_vm_samba_ip`** is used in the following files:
- **cfme/tests/cli/test_appliance_console_db_restore.py**

The *global* fixture **`utility_vm_proxy_data`** is used in the following files:
- **cfme/tests/configure/test_proxy.py**
    - *prepare_proxy_specific*
    - *prepare_proxy_default*

The *global* fixture **`utility_vm_ssh`** is used in the following files:
- **cfme/tests/configure/test_proxy.py**
    - *validate_proxy_logs*
    - *prepare_proxy_specific*
    - *prepare_proxy_default*
    - *test_proxy_valid*
    - *test_proxy_override*

The *local* fixture **`nosleep`** is used in the following files:
- **cfme/utils/tests/test_connect_ssh.py**

The *local* fixture **`vm_mock`** is used in the following files:
- **cfme/utils/tests/test_connect_ssh.py**
    - *test_connect_ssh*

The *local* fixture **`vm_mock2`** is used in the following files:
- **cfme/utils/tests/test_connect_ssh.py**
    - *test_connect_ssh*

Please, consider creating a PRT run to make sure your fixture changes do not break existing usage :smiley:

dajoRH commented:
I detected some fixture changes in commit 0bb7a2a8786759eab4a572f60cf3fb04364c22f8

The *global* fixture **`utility_vm_nfs_ip`** is used in the following files:
- **cfme/tests/cli/test_appliance_console_db_restore.py**

The *global* fixture **`utility_vm_samba_ip`** is used in the following files:
- **cfme/tests/cli/test_appliance_console_db_restore.py**

The *global* fixture **`utility_vm_proxy_data`** is used in the following files:
- **cfme/tests/configure/test_proxy.py**
    - *prepare_proxy_specific*
    - *prepare_proxy_default*

The *global* fixture **`utility_vm_ssh`** is used in the following files:
- **cfme/tests/configure/test_proxy.py**
    - *validate_proxy_logs*
    - *prepare_proxy_specific*
    - *prepare_proxy_default*
    - *test_proxy_valid*
    - *test_proxy_override*

The *local* fixture **`nosleep`** is used in the following files:
- **cfme/utils/tests/test_connect_ssh.py**

The *local* fixture **`vm_mock`** is used in the following files:
- **cfme/utils/tests/test_connect_ssh.py**
    - *test_connect_ssh*

The *local* fixture **`vm_mock2`** is used in the following files:
- **cfme/utils/tests/test_connect_ssh.py**
    - *test_connect_ssh*

Please, consider creating a PRT run to make sure your fixture changes do not break existing usage :smiley:

dajoRH commented:
Would you mind rebasing this Pull Request against latest master, please? :trollface:
*CFME QE Bot*

dajoRH commented:
Lint report for commit f5bbee993495ebc1a0ebbf1dbe6c0c16f972ef17:

`sprout/appliances/models.py`:
- :bangbang: Line 573:51: **E127** *continuation line over-indented for visual indent*

`sprout/sprout/__init__.py`:
- :bangbang: Line 8:1: **E402** *module level import not at top of file*
- :bangbang: Line 9:1: **E402** *module level import not at top of file*
- :bangbang: Line 10:1: **E402** *module level import not at top of file*
- :bangbang: Line 11:1: **E402** *module level import not at top of file*
- :bangbang: Line 12:1: **E402** *module level import not at top of file*

`sprout/sprout/wsgi.py`:
- :bangbang: Line 13:1: **E402** *module level import not at top of file*

![Hmmm](https://i.imgur.com/q3iCWqE.gif)
:scream: You might like to check the lint of your code :point_up: 


**No commit flaws detected**.

*CFME QE Bot*

dajoRH commented:
Lint report for commit 3cd5822c38e54943c1a5c2304215c9996314bae2:

`sprout/appliances/models.py`:
- :bangbang: Line 573:51: **E127** *continuation line over-indented for visual indent*

`sprout/sprout/__init__.py`:
- :bangbang: Line 8:1: **E402** *module level import not at top of file*
- :bangbang: Line 9:1: **E402** *module level import not at top of file*
- :bangbang: Line 10:1: **E402** *module level import not at top of file*
- :bangbang: Line 11:1: **E402** *module level import not at top of file*
- :bangbang: Line 12:1: **E402** *module level import not at top of file*

`sprout/sprout/wsgi.py`:
- :bangbang: Line 13:1: **E402** *module level import not at top of file*

![Hmmm](https://i.imgur.com/q3iCWqE.gif)
:scream: You might like to check the lint of your code :point_up: 


**No commit flaws detected**.

*CFME QE Bot*

dajoRH commented:
I detected some fixture changes in commit 604e4217049f52ae4b93fcfe239fd935d024da95

The *global* fixture **`utility_vm_nfs_ip`** is used in the following files:
- **cfme/tests/cli/test_appliance_console_db_restore.py**

The *global* fixture **`utility_vm_samba_ip`** is used in the following files:
- **cfme/tests/cli/test_appliance_console_db_restore.py**

The *global* fixture **`utility_vm_proxy_data`** is used in the following files:
- **cfme/tests/configure/test_proxy.py**
    - *prepare_proxy_specific*
    - *prepare_proxy_default*

The *global* fixture **`utility_vm_ssh`** is used in the following files:
- **cfme/tests/configure/test_proxy.py**
    - *validate_proxy_logs*
    - *prepare_proxy_specific*
    - *prepare_proxy_default*
    - *test_proxy_valid*
    - *test_proxy_override*

The *local* fixture **`nosleep`** is used in the following files:
- **cfme/utils/tests/test_connect_ssh.py**

The *local* fixture **`vm_mock`** is used in the following files:
- **cfme/utils/tests/test_connect_ssh.py**
    - *test_connect_ssh*

The *local* fixture **`vm_mock2`** is used in the following files:
- **cfme/utils/tests/test_connect_ssh.py**
    - *test_connect_ssh*

Please, consider creating a PRT run to make sure your fixture changes do not break existing usage :smiley:

dajoRH commented:
I detected some fixture changes in commit 604e4217049f52ae4b93fcfe239fd935d024da95

The *global* fixture **`utility_vm_nfs_ip`** is used in the following files:
- **cfme/tests/cli/test_appliance_console_db_restore.py**

The *global* fixture **`utility_vm_samba_ip`** is used in the following files:
- **cfme/tests/cli/test_appliance_console_db_restore.py**

The *global* fixture **`utility_vm_proxy_data`** is used in the following files:
- **cfme/tests/configure/test_proxy.py**
    - *prepare_proxy_specific*
    - *prepare_proxy_default*

The *global* fixture **`utility_vm_ssh`** is used in the following files:
- **cfme/tests/configure/test_proxy.py**
    - *validate_proxy_logs*
    - *prepare_proxy_specific*
    - *prepare_proxy_default*
    - *test_proxy_valid*
    - *test_proxy_override*

The *local* fixture **`nosleep`** is used in the following files:
- **cfme/utils/tests/test_connect_ssh.py**

The *local* fixture **`vm_mock`** is used in the following files:
- **cfme/utils/tests/test_connect_ssh.py**
    - *test_connect_ssh*

The *local* fixture **`vm_mock2`** is used in the following files:
- **cfme/utils/tests/test_connect_ssh.py**
    - *test_connect_ssh*

Please, consider creating a PRT run to make sure your fixture changes do not break existing usage :smiley:

dajoRH commented:
I detected some fixture changes in commit 2b086ffecb66cdc7617c0a1d2a3c69606b53c284

The *global* fixture **`utility_vm_nfs_ip`** is used in the following files:
- **cfme/tests/cli/test_appliance_console_db_restore.py**

The *global* fixture **`utility_vm_samba_ip`** is used in the following files:
- **cfme/tests/cli/test_appliance_console_db_restore.py**

The *global* fixture **`utility_vm_proxy_data`** is used in the following files:
- **cfme/tests/configure/test_proxy.py**
    - *prepare_proxy_specific*
    - *prepare_proxy_default*

The *global* fixture **`utility_vm_ssh`** is used in the following files:
- **cfme/tests/configure/test_proxy.py**
    - *validate_proxy_logs*
    - *prepare_proxy_specific*
    - *prepare_proxy_default*
    - *test_proxy_valid*
    - *test_proxy_override*

The *local* fixture **`nosleep`** is used in the following files:
- **cfme/utils/tests/test_connect_ssh.py**

The *local* fixture **`vm_mock`** is used in the following files:
- **cfme/utils/tests/test_connect_ssh.py**
    - *test_connect_ssh*

The *local* fixture **`vm_mock2`** is used in the following files:
- **cfme/utils/tests/test_connect_ssh.py**
    - *test_connect_ssh*

Please, consider creating a PRT run to make sure your fixture changes do not break existing usage :smiley:

prichard77 commented:
The functionality in this PR has been split into 2 PRs so that additional work can be done on connect_ssh. If this update fixes a test case and doesn't break anything, why wouldn't we push what we have and fix connect_ssh in another PR that will not hold up this original PR (which was intended to fix a test)?

dajoRH commented:
Would you mind rebasing this Pull Request against latest master, please? :trollface:
*CFME QE Bot*

prichard77 commented:
This test is passing but I cannot see the files created via touch on the vm when I check manually. I am not sure this is not a false positive.

prichard77 commented:
I have tested this locally and can see the files created. The sudo command was causing them to get put in a different directory than I was expecting. looks good.
